// @ply-io/ply-sdk/test/utils/mock_app_builder.js
const mockApp = require('./default_app/mock_app');

function MockAppBuilder() {
	this.app = mockApp;

	this.getDefaultApp = function () {
		return mockApp;
	};

	this.setDefaultActionFunction = function (func) {
		this.app.actions.fake_integration.handler = func;
		return this;
	};

	this.disableDefaultActionHttpErrorHandling = function () {
		this.app.actions.fake_integration.disable_http_error_handling = true;
		return this;
	};

	this.addAction = function (action) {
		this.app.actions[action.key] = action;
		return this;
	};

	this.setAuth = function (auth) {
		this.app.auth = auth;
		return this;
	};

	this.disableDefaultAuthHttpErrorHandling = function () {
		this.app.auth.disable_http_error_handling = true;
		return this;
	};

	this.setType = function (type) {
		this.app.type = type;
		return this;
	};

	this.addBeforeMiddleware = function (beforeMiddleware) {
		this.app.beforeMiddlewares.push(beforeMiddleware);
		return this;
	};

	this.addAfterMiddleware = function (afterMiddleware) {
		this.app.afterMiddlewares.push(afterMiddleware);
		return this;
	};

	this.addContextsMapping = function (contextsMapping) {
		this.app.contextsMapping.push(contextsMapping);
		return this;
	};

	this.build = function () {
		return this.app;
	};
}

module.exports = MockAppBuilder;
